Headers/DebugServer2/Architecture/X86/RegisterCopy.h

```diff
@@ -0,0 +1,165 @@
+                user.PASTE(orig_, USER_GP_REG(ax)), 0);                        \
+  } while (0)
+#define DO_COPY_EXTRA_64()                                                     \
+  do {                                                                         \
+    DO_COPY_EXTRA_32();                                                        \
+    DO_COPY_REG(state.linux_gp.fs_base, user.fs_base, 0);                      \
+    DO_COPY_REG(state.linux_gp.gs_base, user.gs_base, 0);                      \
+  } while (0)
+#else
+#define DO_COPY_EXTRA_32()
+#define DO_COPY_EXTRA_64()
+#endif
+
+#if defined(OS_LINUX) && defined(ARCH_X86)
+#define USER_GP_REG(NAME) e##NAME
+#define USER_SS_REG(NAME) x##NAME
+#define USER_FLAGS_REG() eflags
+#elif defined(OS_LINUX) && defined(ARCH_X86_64)
+#define USER_GP_REG(NAME) r##NAME
+#define USER_SS_REG(NAME) NAME
+#define USER_FLAGS_REG() eflags
+#elif defined(OS_FREEBSD) && defined(ARCH_X86_64)
+#define USER_GP_REG(NAME) r_r##NAME
+#define USER_SS_REG(NAME) r_##NAME
+#define USER_FLAGS_REG() r_rflags
+#elif defined(OS_WIN32) && defined(ARCH_X86)
+#define USER_GP_REG(NAME) E##NAME
+#define USER_SS_REG(NAME) Seg##NAME
+#define USER_FLAGS_REG() EFlags
+#endif
+
+#define DO_COPY_GP_REG(NAME)                                                   \
+  DO_COPY_REG(state.gp.STATE_GP_REG(NAME), user.USER_GP_REG(NAME), 0)
+#define DO_COPY_SS_REG(S, U)                                                   \
+  DO_COPY_REG(state.gp.S, user.USER_SS_REG(U), 0xffff)
+#define DO_COPY_FLAGS() DO_COPY_REG(state.gp.eflags, user.USER_FLAGS_REG(), 0)
+
+#define STATE_GP_REG(NAME) e##NAME
+#define DO_COPY_REG(S, U, M) S = (M ? U & M : U)
+template <typename StateType, typename UserStructType>
+static inline void user_to_state32(StateType &state,
+                                   UserStructType const &user) {
+  DO_COPY_GPR_32();
+  DO_COPY_SS();
+  DO_COPY_EXTRA_32();
+}
+#undef DO_COPY_REG
+
+#define DO_COPY_REG(S, U, M) U = (M ? S & M : S)
+template <typename UserStructType, typename StateType>
+static inline void state32_to_user(UserStructType &user,
+                                   StateType const &state) {
+  std::memset(&user, 0, sizeof(user));
+  DO_COPY_GPR_32();
+  DO_COPY_SS();
+  DO_COPY_EXTRA_32();
+}
+#undef DO_COPY_REG
+#undef STATE_GP_REG
+
+#define STATE_GP_REG(NAME) r##NAME
+#define DO_COPY_REG(S, U, M) S = (M ? U & M : U)
+template <typename StateType, typename UserStructType>
+static inline void user_to_state64(StateType &state,
+                                   UserStructType const &user) {
+  DO_COPY_GPR_64();
+  DO_COPY_SS();
+  DO_COPY_EXTRA_64();
+}
+#undef DO_COPY_REG
+
+#define DO_COPY_REG(S, U, M) U = (M ? S & M : S)
+template <typename UserStructType, typename StateType>
+static inline void state64_to_user(UserStructType &user,
+                                   StateType const &state) {
+  std::memset(&user, 0, sizeof(user));
+  DO_COPY_GPR_64();
+  DO_COPY_SS();
+  DO_COPY_EXTRA_64();
+}
```

Comment:
For all that the macros are gross, I think they're probably worth the reduction in code duplication